[PR-1222] Write an article about parser combinators and QQ (#6)
serokell/blog-posts: Haskell Parser Combinators/article.md

## What is a parser combinator and why use them?

<!--
* briefly explain what it is and how it works
* briefly explain their advantages (e.g.: simple to use, idiomatic, a lot of documentation)
* maybe compare with other techniques, like parser generators
  * maybe explain pros and cons
-->

In 2001, Daan Leijen and Erik Meijer published a paper titled [Parsec: Direct Style Monadic Parser Combinators For The Real World](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/parsec-paper-letter.pdf), describing the [`parsec`](https://hackage.haskell.org/package/parsec) library, whose design consequently influenced various others, such as [`megaparsec`](https://hackage.haskell.org/package/megaparsec), [`attoparsec`](https://hackage.haskell.org/package/attoparsec), [`trifecta`](https://hackage.haskell.org/package/trifecta), and even outside the Haskell ecosystem, such as [`NimbleParsec`](https://hexdocs.pm/nimble_parsec/NimbleParsec.html) for Elixir, [`parsec`](https://pythonhosted.org/parsec/) for Python, [`FParsec`](https://www.quanttec.com/fparsec/) for F#, among others.

> Parser combinators are known to be simple to use, without requiring external tools or too many concepts to learn. That is, they are ordinary Haskell functions, that can be combined and returned with other parsers. This in turn also makes them idiomatic to use, being a popular choice among Haskellers, having a great ecosystem, and being easy to find examples and documentation.

Reviewer: The last sentence sounds odd, perhaps you can rephrase it in a simpler way?

Author: Changed, I hope you like it.

3. Many parser combinator libraries will backtrack, which will also be exemplified later, meaning that you might need to be careful to avoid performance penalties.
4. The backtracking mechanism may, in turn, lead to exponential time complexities.

This shouldn't, however, discourage you from their use, as their benefits outweigh the drawbacks. Ultimately, the parsing library of your choice should be chosen based on your needs, and parsing combinators are a good fit for most use cases.

## Implementing a simple parser combinator from scratch

<!--
* make an analogy: there is a tape, and we back move forward (parsed) or
  backward (failed) on this tape.
* datatype to parse String
* satisfy and instances for Monad and Alternative
* maybe explain what Alternative is, since beginners may not be familiar with it
-->

Before we delve into `megaparsec`, it may be interesting to implement our parser combinator from scratch. We will make a simple one that should hopefully help you understand how parser combinators work.

If you instead prefer to jump right into the action, then skip to [Megaparsec tutorial](#megaparsec-tutorial).

### Implementation

> We will begin by importing the appropriate definitions, on a new file called `Parser.hs`:

Reviewer: ```suggestion
We will begin by importing the appropriate definitions in a new file called `Parser.hs`:
```

-- string (x : xs) = (:) <$> char x <*> string xs
-- Or even in terms of monad:
-- string [] = return []
-- string (x : xs) = do
--   y <- char x
--   ys <- string xs
--   return (y : ys)
```

Let's try it in GHCi:

```hs
>>> runParser (string "Haskell") "Haskell"
Right ("Haskell","")

>>> runParser (string "Haskell") "Halloween"
Left [Unexpected 'l']
```

Finally, our last instance: `Alternative`. This may be new for you, so I will explain. First, take a look at its definition, which can be found in `Control.Applicative`:

```hs
class Applicative f => Alternative f where
  empty :: f a
  (<|>) :: f a -> f a -> f a
```

I mentioned that `pure` acts as `1`, which is a parser that always succeeds without consuming input. `empty` works like the converse (or `0`), that is a parser that always fails without consuming input.

For `(<|>)`, we want to implement it such that **the first parser to succeed is used**. That is, we first try the left operand, and failing that, we try the second one. The table below shall guide you:

| |Left parser failed|Left parser succeeded|
|-|-|-|
|Right parser failed|Errors from both parsers|Left parser|
|Right parser succeeded|Right parser|Left parser|

> Here is an implementation of `Alternative`. We constrain `i` and `e` with `Eq` so we can use `nub`, to remove duplicated errors. Note this is not the most efficient or best way to merge errors, but for our purposes, it's good enough.

Reviewer: ```suggestion
Here is an implementation of `Alternative`. We constrain `i` and `e` with `Eq` so we can use `nub` to remove duplicated errors. Note that this is not the best or most efficient way to merge errors, but for our purposes, it's good enough.
```

      | otherwise    -> Left [Error offset $ mkErr hd]
```

If you haven't done the previous exercises, then change `satisfy` in a similar manner, taking `offset` as input and returning `offset + 1`, and error `Error offset $ Unexpected hd` on the error.

For the remainder of the functions, change the parser function to also take `offset` as an argument, and just return it unaffected. Take care to chain these offsets as required. For example, this is the new definition of `Applicative`:

```hs
instance Applicative (Parser i e) where
  pure a = Parser $ \input offset -> Right (offset, a, input)

  Parser f <*> Parser p = Parser $ \input offset ->
    case f input offset of
      Left err -> Left err
      Right (offset', f', rest) ->
        case p rest offset' of
          Left err -> Left err
          Right (offset'', output, rest') -> Right (offset'', f' output, rest')
```

Finally, in `parse`, use `0` as the initial offset.

Keep in mind it's also possible to operate with line and column when you know you have a function operating in characters.

```hs
>>> parse (string "hey") "hi"
Left [Error {erOffset = 1, erError = Expected 'e' 'i'}]
```

> The reader may also choose to instead refactor the parser type to use a custom-defined `State` type that contains the input stream and offset. Many improvements are possible, but we choose to keep simple for didactic purposes.

Reviewer: I would use the word "educational", "didactic" has some negative connotations.

Author: I've never heard about that, but I changed it.

  <hr>
</details>

> ### Foreword

Reviewer: Foreword is an introduction, so I would use a different word like "conclusion", "mid-article conclusion", or something that means that without spelling it out. .

Author: Changed to "conclusion". Seems a bit weird, though...

Reviewer: Hmm, perhaps something like "What's next?"

Author: Cool, I like it! I've changed it.

In this section, we've learned how to make a simple parsing combinator library. There are still various improvements that can be done. To name a few of them, we can implement more instances, such as `Semigroup` and `Monoid`, `IsString`, `MonadFail`, `MonadPlus`, among others. Error messages can be improved by better keeping track of location (such as line and column), supporting more types such as `Text` and `ByteString`, pretty-printing error messages, etc.

We will conclude this part here, and if you are interested, proceed to the next part, where Megaparsec, an industrial-strength parser combinator library will be introduced with a more practical application rather than our simple examples in this section.

## Megaparsec tutorial

<!--
* explain why we chose megaparsec
* parse S-exps
* make a small format for data parsing, no programming constructs
  * equivalent to JSON: objects, lists, null, numbers, booleans, strings
-->

We will now take a look at a practical application with parser combinators: parsing S-expressions. Hopefully, such examples will give you a solid foundation for creating parsers for your grammars.

> Megaparsec's [`README.md`](https://github.com/mrkkrp/megaparsec#comparison-with-other-solutions) offers a nice comparison with other popular parser combinator libraries showing its strengths. Being also pretty common in projects, we will also use it for our tutorial.

Reviewer: ```suggestion
Megaparsec's [`README.md`](https://github.com/mrkkrp/megaparsec#comparison-with-other-solutions) offers a nice comparison with other popular parser combinator libraries showing its strengths.  Since it's pretty common in projects, we will also use it for our tutorial.
```

I would rephrase this and say something like: Megaparsec is the go-to parser for industrial projects, so we will also use it in our tutorial. If you wish to see how it compares to other parser combinator libraries in Haskell, you can read it's README.md 

Otherwise, the start of the paragraph reads a bit confusing.

Author: I've used your suggestion.

  deriving (Show)
```

We also need an alias for the type which `megaparsec` uses.

```hs
type Parser = Parsec
  Void  -- The type for custom error messages. We have none, so use `Void`.
  String  -- The input stream type. Let's use `String` for now, but for better performance, you might want to use `Text` or `ByteString`.
```

Let's start with the simplest of them: `SBool`. We use `<|>` which tries one parser, and failing that, will try the other one.

### Booleans

```hs
bool :: Parser Bool
bool = False <$ string "false" <|> True <$ string "true"
```

We will also make a parser to represent our `SExp`-specific variants. As we progress, we will add more parsers to this list.

Here we use `choice`, which is the same as running `<|>` between each element of the list. `choice [a, b, c] = a <|> b <|> c`.

It's important to notice that `choice`, as well as many other useful combinators, are part of the [`parser-combinators`](https://hackage.haskell.org/package/parser-combinators-1.3.0) package, which Megaparsec conveniently re-exports for us. Should you fail to find a specific function in Megaparsec's documentation, chances are that it's defined in `parser-combinators` instead.

```hs

> atom :: Parser SExp

Reviewer: I had trouble understanding what this one is at start, perhaps it is possible to spell it out more? 

I think, moving the code underneath "We will also make a parser to represent our `SExp`-specific variants. As we progress, we will add more parsers to this list."

and then writing about choice below code sample would be better.

Author: Done! I hope it's better now.

```hs
integer :: Parser Integer
integer = label "integer" $ read <$> (some numberChar <|> ((:) <$> char '-' <*> some numberChar))
```

However, later in the tutorial, we will see an even better way, using Megaparsec's own mechanisms.
  <hr>
</details>

### Strings

To parse a string, the idea is to find everything that is between double-quotes. This approach has a weakness, though: we won't be able to parse escaped quotes. Later in this tutorial, we will present a solution to escaped characters.

To achieve this, we will use two new functions: `between` and `takeWhileP`. Using `between open close x` is equivalent to `open *> x <* close`, while `takeWhileP` will read as many characters as possible that satisfy some predicate.

Note that `takeWhileP` also accepts, optionally, a label for each character that it tries to read. For now, we will leave it as `Nothing`.

```hs
str :: Parser String
str = label "string" $ between (char '"') (char '"') (takeWhileP Nothing (/= '"'))
-- Or:
--- str = label "string" $ char '"' *> takeWhileP Nothing (/= '"') <* char '"'
```

In `atom`, include `SString <$> str`.

```hs
>>> parseTest atom "\"hey\""
SString "hey"
```

### Identifiers

> An identifier will be like a variable in a programming language. We can choose the naming convention for such a thing, but here I chose the ones that are frequently used for C-like languages: the first letter must be a letter or an underscore, while the remainder may be a letter, digit, or underscore.

Reviewer: ```suggestion
An identifier will be like a variable in a programming language. We can choose the naming convention for such a thing, but here I chose the one that is frequently used for C-like languages: the first letter must be a letter or an underscore, while the remainder may be a letter, digit, or underscore.
```

> Here we use the `many` function, which is very similar to the `some` function discussed before. The difference is that `any` tries to run the given parser **0 or more times**. In Regex, our `identifier` parser is equivalent to `[a-zA-Z][a-zA-Z0-9]*`.

Reviewer: Here and elsewhere, I don't think there is a need to capitalize regex.

Author: This was a bit surprising to me, but it makes sense, it's just "regular expressions" which shouldn't be capitalized... I've changed it.

```hs
identifier :: Parser Identifier
identifier = label "identifier" $ do
  first <- letterChar <|> char '_'
  rest <- many $ alphaNumChar <|> char '_'
  pure $ Identifier $ first : rest
-- Or:
-- identifier = Identifier <$> label "identifier" ((:) <$> (letterChar <|> char '_') <*> many (alphaNumChar <|> char '_'))
```

Now include `SId <$> identifier` to `atom`.

**Warning**: When parsing with `<|>` or `choice`, **the order of the parsers matters**. Take this as an example:

```hs
>>> parseTest (SBool <$> bool <|> SId <$> identifier) "true"
SBool True

>>> parseTest (SId <$> identifier <|> SBool <$> bool) "true"
SId (Identifier {getId = "true"})
```

> In the first case, we parse a boolean before an identifier, which causes `true` to be matched by `bool`. However, in the second example, using `identifier` before `bool` caused `true` to be recognized as an identifier. You may assume that parsers are greedy, and you should be careful while considering such things. Notice that in the second case, the `bool` parser will never be executed.

Reviewer: Second sentence is a bit awkward 

You may assume that parsers are greedy, and you should be careful while considering such things.

Such things kind of point to the first clause of the sentence (since they are connected), and it raises confusion since I assume you mean such things = order of parsers. 

Perhaps it can go like this:

You should be careful while considering such things. In most cases, you can assume that parsers are greedy. Notice that in the second case..

Author: Reworded it.

### S-expressions

Finally, let's move on to our last parser, which parses S-expressions themselves, such as `(foo 42 "hey")` or `(bar)`.

First, we will do it the naïve way, which may come as the first natural solution to someone just learning Megaparsec, and then we will do it in a better manner using extra tools offered by Megaparsec.

#### S-expressions: The awkward way

The idea here is to parse an atom, followed by 0 or more atoms. Of course, this means that we want to use `many` to parse such atoms. Additionally, an S-expression may be surrounded by parenthesis, so we can use `between` for such a task.

This parser, however, is not as simple as before, but let's first do it the wrong way (which won't work properly), analyze the problem, and come up with a solution to it.

```hs
sexp :: Parser (SExp, [SExp])
sexp = label "S-expression" $ between (char '(') (char ')') ((,) <$> atom <*> many atom)
-- Or:
-- sexp = label "S-expression" $ char '(' *> ((,) <$> atom <*> many atom) <* char ')'
```

After adding it to `atom` with `uncurry SSExp <$> sexp`, we can test it in GHCi:

```hs
>>> parseTest atom "(foo)"
SSExp (SId (Identifier {getId = "foo"})) []
```

> Ok, looks good so far. What if we have arguments being applied to it?

Reviewer: Isn't it other way around: what if we apply it to arguments?

Author: I don't think it is... I meant we have arguments applied to the function. But I reworded it in any case, hopefully, it's a bit more clear now. Or do we apply a function to its arguments? I'm now confused on what is the correct one :confused:

```
>>> parseTest atom "(foo bar)"
1:5:
  |
1 | (foo bar)
  |     ^
unexpected space
expecting ')', '_', S-expression, alphanumeric character, boolean, identifier, integer, or string
```

Uh-oh. We can't parse spaces, as the error makes it evident. Either we close parenthesis (stop parsing this S-expression), continue the definition of the identifier (`_` or `alphanumeric character`), put another atom (`boolean, identifier, integer, or string`), or another S-expression (by opening parenthesis).

One solution that could come to one's mind is to use the function `sepBy`, which takes a parser and a separator, and use it as ``atom `sepBy` space``. In this case, `space` is a parser defined by Megaparsec that skips zero or more spaces.

```hs
>>> parseTest (between (char '(') (char ')') $ SSExp <$> atom <*> (space *> atom `sepBy` space)) "(foo bar)"
SSExp (SId (Identifier {getId = "foo"})) [SId (Identifier {getId = "bar"})]
```

Of course, this may raise other questions: what if we have a space character before or after the opening and closing parenthesis?

> While that works, the [`megaparsec` documentation](https://hackage.haskell.org/package/megaparsec-9.2.0/docs/Text-Megaparsec-Char-Lexer.html) already offers a solution for this kind of problem, namely, a lexer. Indeed, we will see that some of our previous definitions could have been written more elegantly with the help of this module.

Reviewer: That works = having a space character? Perhaps some rephrasing is necessary.

Author: I meant explicitly skipping spaces. Reworded it.

Begin by defining `skipSpace`. We use the `space` helper function from the lexer module, which expects a function to parse 1 or more spaces, a line comment, and a block comment.

```hs
skipSpace :: Parser ()
skipSpace = L.space
  -- Like `space`, but skips 1 or more space characters.
  space1
  -- Skip from ;; until a newline.
  (L.skipLineComment ";;")
  -- Skip from /* until */. There is also `skipBlockComment`, but it doesn't handle nested comments.
  (L.skipBlockCommentNested "/*" "*/")
```

And now define a `lexeme` parser, which indicates how to consume space after the given parser.

```hs
lexeme :: Parser a -> Parser a
lexeme = L.lexeme skipSpace
```

Now you can use `lexeme` in each of the parsers, and they will automatically skip whitespace after their definitions. Note we also use `lexeme` before our first opening parenthesis, so we can skip space just after the `(`.

```hs
sexp :: Parser (SExp, [SExp])
sexp = label "S-expression" $ lexeme $
  between (lexeme (char '(')) (char ')') ((,) <$> atom <*> many atom)
```

> You may add `lexeme` in other parsers as well, where it's desirable to skip whitespace after that rule. For instance:

Reviewer: ```suggestion
You may add `lexeme` to other parsers where it's desirable to skip whitespace after the rule. For instance:
```

```hs
integer :: Parser Integer
integer = label "integer" $ lexeme $ read <$> some numberChar
```

**Note**: The reader should do the same for `str`, `bool`, and `identifier`.

And now, we can parse S-expressions properly:

```hs
>>> parseTest atom "(  foo  /* 8-) */  bar  ;; look mom, comments! \n  )"
SSExp (SId (Identifier {getId = "foo"})) [SId (Identifier {getId = "bar"})]
```

### Our main parser function

Finally, let's make a function that solves the last two outstanding problems.

For the first one, as mentioned by the Megaparsed documentation, `lexeme` will skip whitespace after a parser, but we also need to deal with whitespace at the beginning of the file.

For the second one, we want to parse everything until we find an `eof` since our parser will currently ignore anything after a complete atom. The two defects are exemplified below:

```hs
>>> parseTest atom "  42"
1:1:
  |
1 |   42
  | ^^^^
unexpected "  42"
expecting S-expression, boolean, identifier, integer, or string

>>> parseTest atom "42 foo bar"
SInteger 42
```

> We will create a function that takes our input string, and either returns an error message or the actual result. Such function "bootstraps" our parser with `between skipSpace eof atom`, which will solve the two defects. It will also serve as an auxiliary function that can be called by client code, rather than `parseTest` which is useful for working on GHCi.

Reviewer: ```suggestion
We will create a function that takes our input string, and either returns an error message or the actual result. Such function "bootstraps" our parser with `between skipSpace eof atom`, which will solve the two defects. It will also serve as an auxiliary function that can be called by client code, rather than `parseTest` which is more useful for working in GHCi.
```

Or some other changes that show contrast, otherwise just saying that it is useful for ghci sounds a bit odd.

In such a situation this is trivial, but care must be taken in more complex examples, for example, what if we wanted to have `(foo bar)` as well as tuples such as `(foo bar, baz quz)`? The S-expression could be very long and lead to exponential complexity in the worst-case scenario. Recently, a very similar problem [was fixed](https://gitlab.com/morley-framework/morley/-/merge_requests/971) in one of our projects.

In this situation, the best thing to do is to _refactor the grammars to share common parts_. The code above could be written without `try` as such:

```hs
numeric :: Parser SExp
numeric = label "number" $ lexeme $ do
  -- Left side before . or f
  left <- some numberChar

  -- .0f or f
  rightM <- optional $ choice
    [ do
        char '.'
        right <- some numberChar
        char' 'f'
        pure $ left ++ "." ++ right
    , do
        char' 'f'
        pure left
    ]

  -- If we had the right side, it's a double, otherwise an integer.
  pure $ case rightM of
    Nothing -> SInteger $ read left
    Just right -> SDouble $ read right
```

And in `atom`, we simply use `numeric` instead of the two previous parsers.

## Rewriting with Megaparsec's lexer

> A lot of the things could be written using the lexer module from Megaparsec. We present the simplified versions below for functions that can be improved.

Reviewer: I would organize it a little bit, at least the 'improvements" should be below the code. 

Because you said "we present the simplified versions below" and I was reading lines like "where's the code, Lebowski?"

Author: Good point, reorganized and reworded this section a bit.

  , quotePat  = dataToPatQ (const Nothing) <=< parseOrThrow
  , quoteType = const $ fail "No S-expression parser for type"
  , quoteDec  = const $ fail "No S-expression parser for declarations"
  }
  where
    parseOrThrow :: String -> Q SExp
    parseOrThrow = either fail pure . parseSExp
```

Additionally, `liftData` and `dataToPatQ` assume that our parsed `SExp` implement the `Data` type class. We will change our `SExp` module to automatically derive this datatype for us. First, enable the `DeriveDataTypeable` language extension, so GHC can automatically implement `Data` for us.

You also need to import `Data` from `Data.Data`.

```hs
{-# LANGUAGE DeriveDataTypeable #-}

module SExp where

import Data.Data (Data)
-- The rest of the file goes here...
```

And now change `Identifier` and `SExp` to derive `Data`:

```hs
newtype Identifier = Identifier
  { getId :: String
  } deriving (Data, Show)

data SExp
  = SSExp    SExp [SExp]  -- (foo 0 "hello" bar), (bar (baz 1)), (foo)
  | SInteger Integer  -- 42
  | SString  String  -- "hello, world"
  | SBool    Bool  -- false, true
  | SId      Identifier  -- foo
  | SDouble  Double  -- 42f, 3.1415f
  deriving (Data, Show)
```

> The basic contents for the `TH` module are summarized as follows:

Reviewer: would link the full file here or somewhere if you have it, since some mistakes with arriving to the final one can appear here, I believe

Author: The full file for the TH module is just above this part, it's just a pretty small file, and the complete code for the parser is linked just before the QuasiQuotation part.

# Parser Combinators in Haskell

<!--
* generic introduction
* say there are four parts for the article
  1. implement custom parser combinator library
  2. megaparsec tutorial
  3. qq

> -->

Reviewer 2: Thanks for the idea of TODO-comments, I'm stealing it ;)

Parser combinators are known to be simple to use without requiring external tools or too many concepts to learn. That is, they are ordinary Haskell functions that can be combined and returned with other parsers. This in turn also makes them idiomatic to use, and being a popular choice among Haskellers, their ecosystem is pretty developed. You should have no trouble finding tutorials and documentation on how to use them.

> It's important to notice that parser combinators also have their flaws. Parser combinators are categorized as LL(∞) parsers, which, elaborating on the points described in Kirill Andreev's [How to Implement an LR(1) Parser](https://serokell.io/blog/how-to-implement-lr1-parser):

Reviewer 2: :blush:

The parsec is _kinda_ LL(1) unless [try](https://hackage.haskell.org/package/parsec-3.1.15.0/docs/Text-Parsec.html#v:try) is used.

Reviewer 2: And, technically, the combinators themselves are _constructors_ of parsers, the idea is not linked to a particular algorithm.

Here is [Earley parser combinator library](https://hackage.haskell.org/package/Earley), for instance.

They are usually implemented as LL(∞), because it is the easiest algorithm of them all.

Author: For the first point, I added a few more discussions about LL(1) and LL(∞) parsers.

For the second point, I reworded this phrase a bit.

1. Don't support left recursion. That is to say, if there is a rule `foo = foo *> bar`, you might need to refactor your grammar otherwise it will loop forever.

> 2. Don't resolve conflicts, which we will see as an example later in this article. This means the ordering of rules can influence the output, and you need to be careful when defining rules that can consume the same input.

Reviewer 2: Or, rather, autoresolve them to left alternative if LL(1) and longest-match-in-a-most-uneffective-way in LL(∞).

Author: I've elaborated a bit more on this point. I hope it's better now.

4. The backtracking mechanism may, in turn, lead to exponential time complexities.

This shouldn't, however, discourage you from their use, as their benefits outweigh the drawbacks. Ultimately, the parsing library of your choice should be chosen based on your needs, and parser combinators are a good fit for most use cases.

## Implementing a simple parser combinator from scratch

<!--
* make an analogy: there is a tape, and we back move forward (parsed) or
  backward (failed) on this tape.
* datatype to parse String
* satisfy and instances for Monad and Alternative
* maybe explain what Alternative is, since beginners may not be familiar with it
-->

Before we delve into `megaparsec`, it may be interesting to implement our parser combinator from scratch. We will make a simple one that should hopefully help you understand how parser combinators work.

If you instead prefer to jump right into the action, then skip to the [Megaparsec tutorial](#megaparsec-tutorial).

### Implementation

We will begin by importing the appropriate definitions in a new file called `Parser.hs`:

```hs
module Parser where

import Control.Applicative (Alternative (..))

> import Data.List (nub)

Reviewer 2: Oh, no.

I understand that this "Set" operation is for sake of simplicity, but eeeeh. I hope it is for `oneOf`, to be used on construction.

Author: Later in the tutorial I describe why we use `nub`, it's just for duplicated errors that may arise in some cases.

```

We will create two data structures that represent our parser.

> The first of them represents some error that may happen while parsing. For now, let's have three errors for now: one that is raised when we are expecting input but there is nothing to consume (`EndOfInput`), one when we find an unexpected character (`Unexpected`), and one for any error messages that the user might want to raise (`CustomError`).

Reviewer 3: Duplicated "for now"

Author: Removed the second "for now".

Our second structure is the actual parser. It's represented as a function that takes some input and either returns a list of errors, if `Left`, or the result of parsing the input together with the rest of the input (that was not parsed), if `Right`.

```hs
data Error i e
  = EndOfInput  -- Expected more input, but there is nothing
  | Unexpected i  -- We didn't expect to find this element
  | CustomError e  -- Extra errors the user may want to create
  | Empty  -- Used in `Alternative` implementation of `empty`
  deriving (Eq, Show)

newtype Parser i e a = Parser

>   { runParser :: [i] -> Either [Error i e] (a, [i])

Reviewer 2: Why not from the transformers? For educational purposes?

Author: Do you mean using `ExceptT` or similar here? I'd prefer to use transformers if we had some blog post on the topic first, but yes, let's keep it simple for educational purposes as well.

  }
```

This is the meaning of each of the type variables:
* `i`: The input stream. For most cases, we will use `Char`, as `[Char]` is the same as `String`. It's interesting to notice that keeping this as a type variable allows us to parse things that are not strings as well.
* `e`: The type of custom error messages. If we don't have those, we may use `Void` instead.
* `a`: The result of our parsing function. It represents the structure parsed from our consumed input.

> Note that we could implement the parser so it also returns the rest of the input even if there were errors, but to keep the implementation a bit simpler, we'll return it only on success.

Reviewer 2: That would be interesting. Do you have any links to the strategies to do that, aside from dropping tokens until we find what we was searching for?

Author: I just meant here that our parser could have type `[i] -> (Either [Error i e] a, [i])`, and the changes would just include the "rest" of the output there (or possibly the input again). But I see now it doesn't look like the best idea. I removed this paragraph.

```hs
instance Monad (Parser i e) where
  return = pure

  Parser p >>= k = Parser $ \input ->
    case p input of
      Left err -> Left err
      Right (output, rest) ->
        let
          Parser p' = k output
        in
        p' rest
  -- Or:
  -- Parser p >>= k = Parser $ \input -> do
  --   (output, rest) <- p input
  --   runParser (k output) rest
```

With our `Applicative` and `Monad` instances, we are now able to **parse sequentially**. For example, we defined `char` above to let us parse a single character, but what if we wanted to parse many of them?

```hs
string :: Eq i => [i] -> Parser i e [i]
string = traverse char
-- Written in a more verbose but perhaps clear way:
-- string :: Eq i => [i] -> Parser i e [i]
-- string [] = pure []
-- string (x : xs) = (:) <$> char x <*> string xs
-- Or even in terms of monad:
-- string [] = return []
-- string (x : xs) = do
--   y <- char x
--   ys <- string xs
--   return (y : ys)

> ```

Reviewer 2: I think, it is worth to write those examples not in comments, but in separate blocks, the article is not in literate haskell format anyway.

Author: I've made separate blocks for them.

      Right (offset', f', rest) ->
        case p rest offset' of
          Left err -> Left err
          Right (offset'', output, rest') -> Right (offset'', f' output, rest')
```

Finally, in `parse`, use `0` as the initial offset.

Keep in mind it's also possible to operate with line and column when you know you have a function operating in characters.

```hs
>>> parse (string "hey") "hi"
Left [Error {erOffset = 1, erError = Expected 'e' 'i'}]
```

The reader may also choose to instead refactor the parser type to use a custom-defined `State` type that contains the input stream and offset. Many improvements are possible, but we choose to keep it simple for educational purposes.
  <hr>
</details>

### Conclusion

In this section, we've learned how to make a simple parsing combinator library. There are still various improvements that can be done. To name a few of them, we can implement more instances, such as `Semigroup` and `Monoid`, `IsString`, `MonadFail`, `MonadPlus`, among others. Error messages can be improved by better keeping track of location (such as line and column), supporting more types such as `Text` and `ByteString`, pretty-printing error messages, etc.

> We will conclude this part here, and if you are interested, proceed to the next part, where Megaparsec, an industrial-strength parser combinator library will be introduced with a more practical application rather than our simple examples in this section.

Reviewer: ```suggestion
We will conclude this part here, and if you are interested, proceed to the next part where Megaparsec, an industrial-strength parser combinator library, will be introduced with a more practical application than our simple examples in this section.
```

Without any further ado, we will now present how to use Megaparsec to parse some [`S-expressions`](https://en.wikipedia.org/wiki/S-expressions).

### Representing S-expressions

To begin, create a file called `SExp.hs`. Also, make sure to get the [`megaparsec`](https://hackage.haskell.org/package/megaparsec) package.

We will use the following data structure to represent various S-expressions. Gradually, we will write a parser that represents each of these cases.

```hs
module SExp where

import Data.Void (Void)
import Text.Megaparsec
import Text.Megaparsec.Char

newtype Identifier = Identifier
  { getId :: String
  } deriving (Show)

data SExp
  = SSExp    SExp [SExp]  -- (foo 0 "hello" bar), (bar (baz 1)), (foo)
  | SInteger Integer  -- 42
  | SString  String  -- "hello, world"
  | SBool    Bool  -- false, true
  | SId      Identifier  -- foo
  deriving (Show)
```

We also need an alias for the type which `megaparsec` uses.

```hs
type Parser = Parsec
  Void  -- The type for custom error messages. We have none, so use `Void`.

>   String  -- The input stream type. Let's use `String` for now, but for better performance, you might want to use `Text` or `ByteString`.

Reviewer 2: I recommend splitting this line on 2 by hand, because otherwise it will be wrapped and wrapper doesn't care about comment indentation.

Reviewer 2: I does go off-window with horizontal scrolling in Github preview.

Author: Good point. I've split them so it's shorter now.

For this parser, we use the `some` function, which tries to run some given parser **1 or more times**. `numberChar` is exported from `Text.Megaparsec.Char` and parses any character that matches a digit. In regex, this is equivalent to `[0-9]+`. This will return us a `[Char]` (aka `String`), which we can `read` to make it into an `Integer`.

```hs
integer :: Parser Integer
integer = read <$> some numberChar
```

We now change our `atom` to also include integers.

```hs
atom :: Parser SExp
atom = choice
  [ SBool <$> bool
  , SInteger <$> integer
  ]
```

Let's also test it in GHCi:

```hs
>>> parseTest atom "42"
SInteger 42

>>> parseTest atom "hey"
1:1:
  |
1 | hey
  | ^^^
unexpected "hey"
expecting "false", "true", or numeric character
```

Although the error message is good, wouldn't it better if it said "integer" instead of "numeric character"? For this, we can use **labels**, which allows us to give a name to a parser. To do this, we can simply change our parser like this:

```hs
integer :: Parser Integer

> integer = label "integer" $ read <$> some numberChar

Reviewer 2: ... or a `<?>` operator with priority of `0`

Author: Right, this also exists. I added a point about it.

```

And now the error message should be slightly improved:

```hs
>>> parseTest atom "hey"
1:1:
  |
1 | hey
  | ^^^
unexpected "hey"
expecting "false", "true" or integer
```

You can also use `label "boolean"` in our `bool` parser to change the message to `expecting boolean or integer`.

**Exercise**: Can you change this parser so it also accepts negative integers?

<details>
  <summary>Solution</summary>
One possible solution using `<|>`:

```hs
integer :: Parser Integer
integer = label "integer" $ read <$> (some numberChar <|> ((:) <$> char '-' <*> some numberChar))
```

However, later in the tutorial, we will see an even better way that uses Megaparsec's own mechanisms.
  <hr>
</details>

### Strings

> To parse a string, the idea is to find everything that is between double-quotes. This approach has a weakness, though: we won't be able to parse escaped quotes. Later in this tutorial, we will present a solution to escaped characters.

Reviewer 2: There is [charLiteral](https://hackage.haskell.org/package/megaparsec-9.2.0/docs/Text-Megaparsec-Char-Lexer.html#v:charLiteral) that parses both single chars and escape sequences like `\n`.

Author: I didn't want to present anything from the `Lexer` module just yet since it's discussed later in the tutorial.

To achieve this, we will use two new functions: `between` and `takeWhileP`. Using `between open close x` is equivalent to `open *> x <* close`, while `takeWhileP` will read as many characters as possible that satisfy some predicate.

Note that `takeWhileP` also accepts, optionally, a label for each character that it tries to read. For now, we will leave it as `Nothing`.

```hs
str :: Parser String

> str = label "string" $ between (char '"') (char '"') (takeWhileP Nothing (/= '"'))

Reviewer 2: ```hs
str = do
  char '"'
  charLiteral `manyTill` char '"'
```

Author: This is how `str` is presented later in the article. But I added another point explaining this part a bit more.

-- Or:
--- str = label "string" $ char '"' *> takeWhileP Nothing (/= '"') <* char '"'
```

In `atom`, include `SString <$> str`.

```hs
>>> parseTest atom "\"hey\""
SString "hey"
```

### Identifiers

An identifier will be like a variable in a programming language. We can choose the naming convention for such a thing, but here I chose the one that is frequently used for C-like languages: the first letter must be a letter or an underscore, while the remainder may be a letter, digit, or underscore.

> Here we use the `many` function, which is very similar to the `some` function discussed before. The difference is that `any` tries to run the given parser **0 or more times**. In regex, our `identifier` parser is equivalent to `[a-zA-Z][a-zA-Z0-9]*`.

Reviewer 3: ```suggestion
Here we use the `many` function, which is very similar to the `some` function discussed before. The difference is that `many` tries to run the given parser **0 or more times**. In regex, our `identifier` parser is equivalent to `[a-zA-Z][a-zA-Z0-9]*`.
```

Reviewer 3: Also, underscore seems to be missing in the regex.

Author: Thanks! I've changed this line.

In other words, make sure that you parse a boolean in `atom` before you parse an identifier.

### S-expressions

Finally, let's move on to our last parser, which parses S-expressions themselves, such as `(foo 42 "hey")` or `(bar)`.

First, we will do it the naïve way, which may come as the first natural solution to someone just learning Megaparsec, and then we will do it in a better manner using extra tools offered by Megaparsec.

#### S-expressions: The awkward way

The idea here is to parse an atom, followed by 0 or more atoms. Of course, this means that we want to use `many` to parse such atoms. Additionally, an S-expression may be surrounded by parenthesis, so we can use `between` for such a task.

This parser, however, is not as simple as before, but let's first do it the wrong way (which won't work properly), analyze the problem, and come up with a solution to it.

```hs
sexp :: Parser (SExp, [SExp])

> sexp = label "S-expression" $ between (char '(') (char ')') ((,) <$> atom <*> many atom)

Reviewer 2: Isn't that `Control.Applicative.some`?

Ah, I get it: nobody made `some` return `Data.List.NotEmpty.NotEmpty` yet.

Author: Actually, there is: https://hackage.haskell.org/package/parser-combinators-1.3.0/docs/Control-Applicative-Combinators-NonEmpty.html#v:some

But Megaparsec doesn't export this function and instead exports the possibly-empty-list one. But I added a small note about it.

```hs
skipSpace :: Parser ()
skipSpace = L.space
  -- Like `space`, but skips 1 or more space characters.
  space1
  -- Skip from ;; until a newline.
  (L.skipLineComment ";;")
  -- Skip from /* until */. There is also `skipBlockComment`, but it doesn't handle nested comments.
  (L.skipBlockCommentNested "/*" "*/")
```

And now define a `lexeme` parser, which indicates how to consume space after the given parser.

```hs
lexeme :: Parser a -> Parser a
lexeme = L.lexeme skipSpace
```

Now you can use `lexeme` in each of the parsers, and they will automatically skip whitespace after their definitions. Note that we also use `lexeme` before our first opening parenthesis, so we can skip space just after the `(`.

```hs
sexp :: Parser (SExp, [SExp])
sexp = label "S-expression" $ lexeme $
  between (lexeme (char '(')) (char ')') ((,) <$> atom <*> many atom)
```

You may add `lexeme` to other parsers where it's desirable to skip whitespace after the rule. For instance:

```hs
integer :: Parser Integer
integer = label "integer" $ lexeme $ read <$> some numberChar
```

> **Note**: The reader should do the same for `str`, `bool`, and `identifier`.

Reviewer 2: ... or on whole atom instead, to not repeat themselves.

Author: Ah, nice observation! I didn't think of adding it to `atom`. I added a comment about it.

  * We use our standard Megaparsec parser, `parseSExp`, but fail if it can't parse (halting compilation). `liftData` then transforms our `SExp` into an `Exp`.
  * If this is your first time seeing `<=<`, it's just another way to compose monadic functions. `g <=< f` is the same as `\x -> g =<< f x`.
  * If you've read our previous TH tutorial, it creates an `Exp` just like `[e||]`.
* `quotePat` creates a Haskell pattern.
  * Example: `case mySExp of {[sexp|(foo 42)|] -> True; _ -> False}`.
  * Similarly to `quoteExp`, we use `dataToPatQ (const Nothing)` to transform our `SExp` into a `Pat`.
    * Confusingly, `liftData` is defined as `dataToExpQ (const Nothing)`.
    * The `const Nothing` can be replaced to handle type-specific cases. We are fine with the default behavior for now but we'll change it later.
  * Creates a `Pat`, like `[p||]`.
* `quoteType` creates a Haskell type.
  * Since we have no `SExp`, `SInteger` etc types matching our constructors, we will fail when someone tries to use our parser as types.
  * Creates a `Type`, like `[t||]`.
* `quoteDec` creates a Haskell declaration.
  * Similarly to `quoteType`, we won't define this one.

>   * Creates a `Pat`, like `[p||]`.

Reviewer 3: Copy-paste mistake?

Author: Good catch, it was a copy-paste mistake. I fixed it.

```hs
  , quotePat  = dataToPatQ antiQuoteVar <=< parseOrThrow
```

Now check whether it works:

```hs
>>> case [sexp|(foo 42)|] of
...  [sexp|(foo $answer)|] -> print answer
...  _ -> putStrLn "oops"
SInteger 42
```

It's also possible to give a meaning to `$answer` at the expression level by having it capture a Haskell variable instead. Let's change `TH` a bit to support it:

```hs
    antiQuoteVar :: Typeable a => (Name -> Q b) -> a -> Maybe (Q b)
    antiQuoteVar mkVar var = case cast var of
      Nothing -> Nothing
      Just var' -> case var' of
        STH var'' -> Just $ mkVar $ mkName var''
        _other    -> Nothing
```

And our `QuasiQuoter` now has:

```hs
  { quoteExp  = dataToExpQ (antiQuoteVar varE) <=< parseOrThrow
  , quotePat  = dataToPatQ (antiQuoteVar varP) <=< parseOrThrow
```

Don't forget to import `dataToExpQ`, which now replaces `liftData` as well as `varE`. We can now test our quasi-quoter again:

```hs
>>> let x = SInteger 42 in case [sexp|(foo $x)|] of
          [sexp|(foo $answer)|] -> print answer
          _ -> putStrLn "oops"
SInteger 42
```

And that's it! You now support quasi-quotations.

> The complete code for the parser and quasi-quoter may be found [here](https://gist.github.com/heitor-lassarote/f784c16da80bcfc5271558ede51c70d9).

Reviewer 3: Wow, those two tricks are really cool, I severely lacked them during my development.

Our second structure is the actual parser. It's represented as a function that takes some input and either returns a list of errors, if `Left`, or the result of parsing the input together with the rest of the input (that was not parsed), if `Right`.

```hs
data Error i e
  = EndOfInput  -- Expected more input, but there is nothing
  | Unexpected i  -- We didn't expect to find this element
  | CustomError e  -- Extra errors the user may want to create
  | Empty  -- Used in `Alternative` implementation of `empty`
  deriving (Eq, Show)

newtype Parser i e a = Parser
  { runParser :: [i] -> Either [Error i e] (a, [i])
  }
```

This is the meaning of each of the type variables:
* `i`: The input stream. For most cases, we will use `Char`, as `[Char]` is the same as `String`. It's interesting to notice that keeping this as a type variable allows us to parse things that are not strings as well.
* `e`: The type of custom error messages. If we don't have those, we may use `Void` instead.
* `a`: The result of our parsing function. It represents the structure parsed from our consumed input.

> Let's begin by creating the most primitive function: `satisfy`. It takes some predicate that operates on the current character, and advances the parser, if possible, returning the consumed character.

Reviewer: Is it possible to have a bit simpler structure here?

takes a predicate that operates on the current character, advances the parser (if possible), and returns the consumed character.

Author: I reworded a bit, maybe it's easier to understand now.

```hs
char :: Eq i => i -> Parser i e i
char i = satisfy (== i)
```

If you load this in GHCi, you should now be able to test it:
```hs
>>> :l Parser

>>> runParser (char 'h') "hello"
Right ('h', "ello")

>>> runParser (char 'h') "greetings"
Left [Unexpected 'g']

>>> runParser (char 'h') ""
Left [EndOfInput]
```

Let's now create `Monad` and `Alternative` instances for our parser. For educative purposes, we will also explicitly implement `Functor` and `Applicative` instead of `fmap = liftM` and `(<*>) = ap`.

We start with `Functor`. Since we map over the `a` type variable obtained in the case of a successful parse, we can only apply `f` in case if we get `Right`.

```hs
instance Functor (Parser i e) where
  fmap f (Parser p) = Parser $ \input ->
    case p input of
      Left err -> Left err
      Right (output, rest) -> Right (f output, rest)
```

Here's a little shorter implementation. Since `Either` is a monad, we can use do-notation.

```hs
instance Functor (Parser i e) where
  fmap f (Parser p) = Parser $ \input -> do
    (output, rest) <- p input
    pure (f output, rest)
```

> It's interesting to notice that `Either` is a monad, so we can also write the definition of most things more concisely, as shown in the commented part.

Reviewer: You mentioned that in the previous line already, sounds a bit awkward now. :D (+ I assume the reference to the "commented" part is now invalid)

Author: Some leftover, I removed this part.

    case l input of
      Left err ->
        case r input of
          Left err' -> Left $ nub $ err <> err'
          Right (output, rest) -> Right (output, rest)
      Right (output, rest) -> Right (output, rest)
```

Finally, let's test it!

```hs
>>> runParser (string "hello" <|> string "greetings") "hello, world"
Right ("hello",", world")

>>> runParser (string "hello" <|> string "greetings") "greetings, world"
Right ("greetings",", world")

>>> runParser (string "hello" <|> string "greetings") "bye, world"
Left [Unexpected 'b']

>>> runParser (empty <|> pure ()) ""
Right ()

>>> runParser (pure () <|> empty) ""
Right ()
```

We can see our errors composing with a more complicated expression:

```hs
>>> runParser ((string "hello" *> string ", globe") <|> string "greetings") "hello, world"
Left [Unexpected 'w',Unexpected 'h']
```

Hopefully, this should give you an intuition of how parser combinators work. Popular parser combinator libraries are more complex than this, as they tend to keep track of extra state for better error messages, optimization, information about line and column, etc., but the overall operations between them it similar.

> The complete code for the parser with exercises solutions can be found [here](https://gist.github.com/heitor-lassarote/3e7314956e86b8227f6f6040e69aca9d).

Reviewer: There is something wrong here: 

1. exercises & solutions? 
2. I only see parser code in the gist.

Author: I reworded a bit, mentioned it contains the solutions to the exercises below.

> I only see parser code in the gist.

I mentioned it's only for this section, which only has the parser code. Or am I missing something?

Reviewer: I meant that it doesn't have the solutions to the exercises, at least it seems that way to me.

Reviewer: My bad, I thought it didn't have the solutions, but it does.

## Megaparsec tutorial

<!--
* explain why we chose megaparsec
* parse S-exps
* make a small format for data parsing, no programming constructs
  * equivalent to JSON: objects, lists, null, numbers, booleans, strings
-->

We will now take a look at a practical application with parser combinators: parsing S-expressions. Hopefully, such examples will give you a solid foundation for creating parsers for your grammars.

Megaparsec is the go-to parser for industrial projects, so we will also use it in our tutorial. If you wish to see how it compares to other parser combinator libraries in Haskell, you can read its [`README.md`](https://github.com/mrkkrp/megaparsec#comparison-with-other-solutions).

Keep in mind you should choose the parser library that better suits your needs. If you plan to parse machine-written data, for example, then Attoparsec may be a better choice. Its interface is similar to Megaparsec's, so the knowledge you will learn here will also be useful.

> Without any further ado, we will now present how to use Megaparsec to parse some [`S-expressions`](https://en.wikipedia.org/wiki/S-expressions).

Reviewer: I would link in the first paragraph, since people will probably Google when seeing the first occurance. Also, I don't think it needs the code-font.

Author: I linked in the first paragraph and removed the code-font.

```hs
bool :: Parser Bool
bool = False <$ string "false" <|> True <$ string "true"
```

We will also make a parser to represent our `SExp`-specific variants. As we progress, we will add more parsers to this list.

```hs
atom :: Parser SExp
atom = choice
  [ SBool <$> bool
  ]
```

Here we use `choice`, which is the same as running `<|>` between each element of the list. `choice [a, b, c] = a <|> b <|> c`.

It's important to notice that `choice`, as well as many other useful combinators, are part of the [`parser-combinators`](https://hackage.haskell.org/package/parser-combinators-1.3.0) package, which Megaparsec conveniently re-exports for us. Should you fail to find a specific function in Megaparsec's documentation, chances are that it's defined in `parser-combinators` instead.

Now, fire up GHCi and test that we can parse it! `megaparsec` provides the `parseTest` function which allows us to conveniently test our parsers.

```hs
>>> parseTest atom "false"
SBool False

>>> parseTest atom "hey"
1:1:
  |
1 | hey
  | ^^^
unexpected "hey"
expecting "false" or "true"
```

### Integers

> To parse integers, one strategy is to parse as many digits (chars whose value is between `'0'` and `'9'`) as possible. This will give us a list of characters, which we can then `read`,

Reviewer: Something missing here? Or perhaps we just need to change colon to dot. :D

Author: Eagle eyes, I read this two times and didn't spot it. But now I changed it. Thanks!
